src/Agent.Worker/TaskDefinitionManager.cs

```diff
@@ -0,0 +1,291 @@
+            string json = File.ReadAllText(file);
+            definition.Data = JsonConvert.DeserializeObject<DefinitionData>(json);
+
+            // Replace the macros within the handler data sections.
+            foreach (HandlerData handlerData in (definition.Data?.Execution?.All as IEnumerable<HandlerData> ?? new HandlerData[0]))
+            {
+                handlerData?.ReplaceMacros(definition);
+            }
+
+            return definition;
+        }
+    }
+
+    public sealed class Definition
+    {
+        public DefinitionData Data { get; set; }
+        public string Directory { get; set; }
+    }
+
+    public sealed class DefinitionData
+    {
+        public TaskInputDefinition[] Inputs { get; set; }
+        public ExecutionData Execution { get; set; }
+    }
+
+    public sealed class ExecutionData
+    {
+        private readonly List<HandlerData> _all = new List<HandlerData>();
+        private NodeHandlerData _node;
+        private ProcessHandlerData _process;
+
+        [JsonIgnore]
+        public List<HandlerData> All => _all;
+
+        public NodeHandlerData Node
+        {
+            get
+            {
+                return _node;
+            }
+
+            set
+            {
+                _node = value;
+                Add(value);
+            }
+        }
+
+#if !OS_WINDOWS
+        [JsonIgnore]
+#endif
+        public ProcessHandlerData Process
+        {
+            get
+            {
+                return _process;
+            }
+
+            set
+            {
+                _process = value;
+                Add(value);
+            }
+        }
+
+        private void Add(HandlerData data)
+        {
+            if (data != null)
+            {
+                _all.Add(data);
+            }
+        }
+    }
+
+    public abstract class HandlerData
+    {
+        // TODO: Revisit whether we want to handle this variable differently.
+        // TODO: Move to constants.
+        public static readonly string CurrentDirectoryMacro = "$(currentdirectory)";
+
+        public string[] Platforms { get; set; }
+        [JsonIgnore]
+        public abstract int Priority { get; }
+        public string Target { get; set; }
+
+        public bool MatchesPlatform()
+        {
+#if OS_WINDOWS
+            const string CurrentPlatform = "windows";
+#else
+            const string CurrentPlatform = ""; // TODO: What to do here? Update test when changed.
+            if (string.IsNullOrEmpty(CurrentPlatform))
+            {
+                return false;
+            }
+#endif
+
+            return Platforms?.Any(x => string.Equals(x, CurrentPlatform, StringComparison.OrdinalIgnoreCase)) ?? false;
+        }
+
+        public virtual void ReplaceMacros(Definition definition)
+        {
+            Target = Replace(input: Target, macro: CurrentDirectoryMacro, replacement: definition.Directory);
+        }
+
+        protected static string Replace(string input, string macro, string replacement)
```

Comment:
When I add the recursive variable expansion for the job variables, I'll see if there is a better (common) home for this logic.